Multiple buttons

cs-shadow opened the pull request:
d.o issue: https://www.drupal.org/node/2273361

Note: This is just the first step in order to support different types of entities for selection. Read my last comment (short essay: https://www.drupal.org/node/2273361#comment-8999597) on the issue for more details.

cs-shadow committed:
Add option to configure entity specific buttons

cs-shadow committed:
Add buttons for the selected entities in the toolbar configuration

cs-shadow committed:
Add commands and buttons for specific entities

slashrsm commented on config/schema/entity_embed.schema.yml:
```diff
@@ -0,0 +1,14 @@
+      label: 'Whether or not to enable the genric embed button'
+    entity_types:
+      type: sequence
+      label: 'Entities for which embed button is to be enabled'
```
Entity types

slashrsm commented on entity_embed.module:
```diff
@@ -0,0 +1,29 @@
+function entity_embed_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    case 'entity_embed.admin':
+      return '<p>' . t("By default, only generic entity embed button is enabled. If you want to enable entity embed buttons for specific entity types, you need to select the desired entities. Once you've enabled the desired buttons, those buttons will be available in the Toolbar Configuration UI. To enable those buttons for the CKEditor user interface, go to <a href='@filterformats'>Text formats  ...
```
Only generic entity embed button is enabled by default. You can enable entity type specific buttons on the <a href="@settings_page">settings page</a>. When you enabled them go to <a href='@filterformats'>Text formats and editors</a> configuration page to add them to CKEditor's toolbar.

cs-shadow replied:
There's no need for the 'settings page' link , since this text is for the settings page only.

slashrsm replied:
Ah :) OK.

slashrsm commented on entity_embed.module:
```diff
@@ -0,0 +1,29 @@
+function entity_embed_permission() {
+  return array(
+    'administer entity embed settings' => array(
+      'title' => t('Administer entity embed settings'),
```
"Administer entity embed" and "administer entity embed"

slashrsm commented on js/plugins/drupalentity/plugin.js:
```diff
@@ -96,13 +96,26 @@
+          editor.ui.addButton(button, {
+            label: Drupal.t(buttons[button].label),
+            command: command,
+            icon: this.path + '/entity.png',
```
Make this configurable.

cs-shadow replied:
Yes, that's on my list of todo's. Since I'm still not sure how to allow user to config icons, I plan to do this as follow-up. I wanted to get a basic structure in first.

slashrsm replied:
OK, follow-up is OK.

slashrsm commented on src/Form/EntityEmbedSettingsForm.php:
```diff
@@ -0,0 +1,63 @@
+
+    $form['generic_form'] = array(
+      '#type' => 'checkbox',
+      '#title' => $this->t('Enable the genric embed button'),
```
Add #description.

slashrsm commented on src/Form/EntityEmbedSettingsForm.php:
```diff
@@ -0,0 +1,63 @@
+      '#default_value' => $config->get('enable_generic_form'),
+    );
+    $form['entity_types'] = array(
+      '#title' => $this->t('Select entities for which specific embed button is to be enabled'),
```
Entity types. Make this shorter and #description for longer instructions.

slashrsm commented on src/Plugin/CKEditorPlugin/DrupalEntity.php:
```diff
@@ -20,17 +22,37 @@
   public function getButtons() {
-    return array(
-      'DrupalEntity' => array(
+    $config = \Drupal::config('entity_embed.settings');
```
Can we inject this?

cs-shadow committed:
Fix help text and inject config_factory object

cs-shadow commented:
@slashrsm, Fixed the concerns you raised and renamed `enabel_generic_form` to `enable_generic_button` since that reflects better its meaning. Let me know if anything needs to be changed.

slashrsm commented:
Looks good to me.

jibran commented on src/Form/EntityEmbedSettingsForm.php:
```diff
@@ -0,0 +1,64 @@
+ * Configure entity_embed settings for this site.
+ */
+class EntityEmbedSettingsForm extends ConfigFormBase {
+  use EntityHelperTrait;
```
I don't think this is needed for `$this->entityManager()` you can remove it and inject the service. It'll be good for unit testing.

davereid commented:
I'm feeling like this is the wrong path forward. From the drupal.org issue:

> I'm feeling that entity embed should have an 'Embed button configuration entity'. It would be easy to supply individual button icons then, since it would be a file upload on the config entity? And then modules like Media could ship with it's own config button entity and a known ID that it can alter.

davereid commented:
After talking with @cs-shadow today we're going to close this without merging and retry another approach using configuration entities for the buttons so that modules can ship with install config entities for pre-defined buttons.
